source/game/system/CourseMap.cc

```diff
@@ -233,6 +233,372 @@ CourseMap::~CourseMap() {
+            for (size_t j = 0; j < next->prevCount(); ++j) {
+                MapdataCheckPoint *prev = next->prevPoint(j);
+
+                if (prev == checkpoint) {
+                    continue;
+                }
+
+                id = findRecursiveSector(pos, 1, true, prev, distanceRatio, false);
+                if (id != -1) {
+                    return id;
+                }
+            }
+        }
+
+        // Step 3: If step 2 fails, start at previous checkpoint(s) and search forwards
+        for (size_t i = 0; i < checkpoint->prevCount(); ++i) {
+            MapdataCheckPoint *prev = checkpoint->prevPoint(i);
+
+            for (size_t j = 0; j < prev->nextCount(); ++j) {
+                MapdataCheckPoint *next = prev->nextPoint(j);
+
+                if (next == checkpoint) {
+                    continue;
+                }
+
+                id = findRecursiveSector(pos, 1, false, next, distanceRatio, false);
+                if (id != -1) {
+                    return id;
+                }
+            }
+        }
+
+        // Step 4: If step 3 fails, start at current checkpoint and search backwards
+        return searchPrevCheckpoint(pos, 0, checkpoint, distanceRatio, false, false);
+    } else {
+        // (do all the steps from 2a in reverse)
+        // Search prev
+        id = searchPrevCheckpoint(pos, 0, checkpoint, distanceRatio, false, false);
+
+        if (id != -1) {
+            return id;
+        }
+
+        // If that fails, search prev -> next
+        for (size_t i = 0; i < checkpoint->prevCount(); ++i) {
+            MapdataCheckPoint *prev = checkpoint->prevPoint(i);
+
+            for (size_t j = 0; j < prev->nextCount(); ++j) {
+                MapdataCheckPoint *next = prev->nextPoint(j);
+
+                if (next == checkpoint) {
+                    continue;
+                }
+
+                id = findRecursiveSector(pos, 1, false, next, distanceRatio, false);
+
+                if (id != -1) {
+                    return id;
+                }
+            }
+        }
+
+        // If that fails, search next -> prev
+        for (size_t i = 0; i < checkpoint->nextCount(); ++i) {
+            MapdataCheckPoint *next = checkpoint->nextPoint(i);
+
+            for (size_t j = 0; j < next->prevCount(); ++j) {
+                MapdataCheckPoint *prev = next->prevPoint(j);
+
+                if (prev == checkpoint) {
+                    continue;
+                }
+
+                id = findRecursiveSector(pos, 1, true, prev, distanceRatio, false);
+
+                if (id != -1) {
+                    return id;
+                }
+            }
+        }
+
+        // If that fails, search next
+        return searchNextCheckpoint(pos, 0, checkpoint, distanceRatio, false, false);
+    }
+
+    return id;
+}
+
+s16 CourseMap::findSectorOutsideSector(const EGG::Vector3f &pos, MapdataCheckPoint *checkpoint,
+        f32 &distanceRatio) {
+    s16 id = -1;
+
+    // Search next -> prev
```

Comment:
```suggestion
    // Step 1: Starting at next checkpoint(s), search backwards
```
The other comments following this should be consistent.